--- a/base/command_line.h
+++ b/base/command_line.h
@@ -19,6 +19,7 @@
 #include <stddef.h>
 #include <functional>
 #include <map>
+#include <memory>
 #include <string>
 #include <vector>
 
--- a/ui/ozone/platform/wayland/gpu/drm_render_node_path_finder.cc
+++ b/ui/ozone/platform/wayland/gpu/drm_render_node_path_finder.cc
@@ -4,6 +4,7 @@
 
 #include "ui/ozone/platform/wayland/gpu/drm_render_node_path_finder.h"
 
+#include <cstring>
 #include <fcntl.h>
 #include <sys/stat.h>
 #include <sys/types.h>
--- a/chrome/browser/new_tab_page/modules/safe_browsing/safe_browsing_handler.cc
+++ b/chrome/browser/new_tab_page/modules/safe_browsing/safe_browsing_handler.cc
@@ -13,10 +13,8 @@
 #include "components/prefs/pref_service.h"
 #include "components/safe_browsing/core/browser/safe_browsing_metrics_collector.h"
 #include "components/safe_browsing/core/common/safe_browsing_policy_handler.h"
-#include "components/safe_browsing/core/common/safe_browsing_prefs.h"
 #include "components/search/ntp_features.h"
 
-using ::safe_browsing::IsEnhancedProtectionEnabled;
 using ::safe_browsing::SafeBrowsingMetricsCollectorFactory;
 using ::safe_browsing::SafeBrowsingPolicyHandler;
 
@@ -66,7 +64,7 @@
   bool managed =
       SafeBrowsingPolicyHandler::IsSafeBrowsingProtectionLevelSetByPolicy(
           pref_service_);
-  bool already_enabled = IsEnhancedProtectionEnabled(*pref_service_);
+  bool already_enabled = false;
   bool module_already_opened =
       pref_service_->GetBoolean(prefs::kSafeBrowsingModuleOpened);
   base::Time cooldown_end = GetLastCooldownEndTime(pref_service_);
